### ISSUE TITLE
Clarify how associative arrays as document values should be updated.

### ISSUE THREAD

**Issue by internalsystemerror** [event-engine/php-document-store#17]
Assuming the following:
```
        $this->documentStore->updateDoc(
            $collection,
            $event->get(Payload::USER_ID),
            [
                Payload::USER_VALIDATIONS   => [
                    $event->get(Payload::USER_EMAIL_ADDRESS) => $event->createdAt(),
                ],
            ],
        );
```

The `InMemoryDocumentStore` would merge the `Payload::USER_VALIDATIONS` arrays recursively, adding additional values. This provides no way of removing a nested key. There is even a test to enforce this behaviour: https://github.com/event-engine/php-document-store/blob/1e980c8e8e3f7b48428740e109692440dffa3a2b/tests/InMemoryDocumentStoreTest.php#L88-112

Further, the `event-engine/php-postgres-document-store` handles this differently, where merges are NOT recursive: https://github.com/event-engine/php-postgres-document-store/blob/e9721e21f42a648a79eb9865b5656c6949bf1c42/src/PostgresDocumentStore.php#L366 

For more on how the `||` operator is used: https://www.postgresql.org/docs/13/functions-json.html

I have submitted the following PRs for changes:

- Add tests to `event-engine/php-postgres-document-store` to confirm that associative arrays are not merged recursively (https://github.com/event-engine/php-postgres-document-store/pull/15).
- Modify the tests on this repository, and the behaviour of `InMemoryDocumentStore` to conform to the behaviour of `event-engine/php-postgres-document-store` (https://github.com/event-engine/php-document-store/pull/18).

**Comment by codeliner** [comment]
@sandrokeil @martin-schilling @heiglandreas @jsor @sblum I consider this a bugfix and would release a `v0.6.1` on next Tuesday. What do you think? It might effect some test cases, but that's even a good thing, because it won't work with Postgres anyway.

**Comment by codeliner** [comment]
Thank you for investigation and the PRs @internalsystemerror :clap:

**Comment by martin-schilling** [comment]
Just for completeness sake, I think I still remember the reason `arrayReplaceRecursiveAssocOnly` was added. In a previous [commit](https://github.com/event-engine/php-document-store/commit/308ebf11ce74d993d3954aebdad555375bb399f2#diff-ab34fe25b5d11dbd29947e5c6a4b980dL137-R175), `array_merge` had been replaced by `array_replace_recursive` which led to changed behavior and failing tests. The `arrayReplaceRecursiveAssocOnly` introduced [here](https://github.com/event-engine/php-document-store/commit/e02c1140fd5e852d3dd5e7b3abacea6ae3efd0b0#diff-ab34fe25b5d11dbd29947e5c6a4b980d) was a compromise between the two methods. Working with PHP arrays can really be a mess.  
Anyway, I'm not entirely sure why we didn't notice that Postgres behaves differently here but I feel like the InMemoryDocumentStore should behave similarly to the PostgresDocumentStore. 

Just out of curiosity though, how do other DBs behave? It might be worth checking how this works in other popular DBs like MongoDB to make sure that the most suitable procedure can be chosen. Afterwards this behavior should probably be defined and documented so that we don't have this problem in the future anymore.

Thanks for the PR!

**Comment by codeliner** [comment]
Good points @martin-schilling . I ask @sandrokeil tomorrow regarding MongoDB.

**Comment by codeliner** [comment]
MongoDB has a [$set](https://docs.mongodb.com/manual/reference/operator/update/set/#up._S_set) operator and it behaves similar to Postgres. I'm going to merge the PRs now.

**Comment by codeliner** [comment]
Fixed with #18
